ISPC: unblock 1.17-1.19

heerener opened the pull request:
With these changes, I've managed to build & install the three latest ISPC versions

heerener committed:
ispc: attempts at getting more recent versions to work

heerener committed:
ispc: more attempts to get newer versions to build

heerener committed:
ispc: cleanup

heerener committed:
llvm: remove ispc_patches variant again

heerener committed:
ispc: unpin ncurses

heerener committed:
ispc: satisfy style checks

tldahlgren reviewed (commented):
Confirmed the sha256 but am unable to confirm the build due to `llvm` build failures I'm having.

kwryankrattiger commented on var/spack/repos/builtin/packages/ispc/package.py:
```diff
@@ -60,6 +61,11 @@ class Ispc(CMakePackage):
     )
 
+    def setup_build_environment(self, env):
+        if self.spec.version >= Version("1.18.0"):
```
I think the spackism for this is `self.spec.satisfies("@1.18.0:")`, but I don't see why this doesn't also work.

@tldahlgren is it worth changing? I am not sure if there are other details considered in the `spec.satisfies` that would be  missed by this version comparison.

tldahlgren replied:
I agree, in general it is a good habit to get into when checking the package's spec to use `self.spec.satisfies()`.

(More on versions at https://spack.readthedocs.io/en/latest/packaging_guide.html#version-comparison).

aumuell replied:
Thank you for putting in the work to build this!

In the meantime, LLVM 16 seems to have hit develop, but this breaks the build for ispc, as it is only compatible with LLVM 13-15. Because of this, I would like to add at least https://github.com/aumuell/spack/commit/79432e2830de49cce6f70fc0bb001720d9f9cefd to this PR. Unfortunately, I do not know how to do that. (I tried pushing to your ispc branch and also could not find how to make a PR to your ispc branch).

If you are ok with including my additional changes from https://github.com/aumuell/spack/tree/heerener-ispc, please pull them into your branch. But I think that this should get merged either way.

heerener replied:
Thanks for the clarifications! I've managed to create the PR by starting from your fork.
Where did you find what versions of llvm ispc requires? Because the version restrictions I set were mostly guesswork.

aumuell replied:
> Where did you find what versions of llvm ispc requires? Because the version restrictions I set were mostly guesswork.

Unfortunately I don't know where this information can be looked up. That 1.19 is only compatible with LLVM 13-15 was printed to me while it was trying to build against LLVM 16. For other versions of ispc I don't know either.

heerener replied:
Ah! I've seen that same message for one of the versions, just didn't think to make it fail on purpose for the others :-)

aumuell committed:
ispc: checksum 1.18.1

just in order to include all versions, also checked that it builds

aumuell committed:
ispc: 1.19 is only compatible with LLVM 13-15

otherwise it would not build against develop, as this now has LLVM 16

aumuell committed:
ispc: use spec.satisfies instead of version comparison

according to suggestions from review and docs, this is the canonical way to do it

aumuell committed:
ispc: relax LLVM version to what ispc requires itself

verified that it builds against LLVM 13, 14, 15, but not 12 and 16

aumuell reviewed (changes requested):
please limit to LLVM < 16, as ispc@1.19 only builds against LLVM 13-15 (see other comment for patch)

heerener committed:
Merge pull request #1 from aumuell/heerener-ispc

Heerener ispc

aumuell reviewed (approved):
LGTM: this worked nicely with ISPC 1.18, 1.19 and LLVM 13-15 and ispc could be used to compile embree

Thank you!

tldahlgren reviewed (approved):
LGTM
